drivers/sensor/icm42605/icm42605.c

```diff
@@ -0,0 +1,467 @@
+					+ (drv_data->fifo_data[4] & 0xff);
+				drv_data->accel_z =
+					((drv_data->fifo_data[5] & 0xff) << 8)
+					+ (drv_data->fifo_data[6] & 0xff);
+			}
+			if (!(drv_data->fifo_data[0] & BIT_FIFO_HEAD_GYRO)) {
+				drv_data->temp =
+					(int16_t)(drv_data->fifo_data[7]);
+			} else {
+				if (!(drv_data->fifo_data[7] ==
+				      FIFO_GYRO0_RESET_VALUE &&
+				      drv_data->fifo_data[8] ==
+				      FIFO_GYRO1_RESET_VALUE)) {
+					drv_data->gyro_x =
+						((drv_data->fifo_data[7] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[8] &
+						   0xff);
+					drv_data->gyro_y =
+						((drv_data->fifo_data[9] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[10] &
+						   0xff);
+					drv_data->gyro_z =
+						((drv_data->fifo_data[11] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[12] &
+						   0xff);
+				}
+				drv_data->temp =
+					(int16_t)(drv_data->fifo_data[13]);
+			}
+		} else {
+			if (drv_data->fifo_data[0] & BIT_FIFO_HEAD_GYRO) {
+				if (!(drv_data->fifo_data[1] ==
+				      FIFO_GYRO0_RESET_VALUE &&
+				      drv_data->fifo_data[2] ==
+				      FIFO_GYRO1_RESET_VALUE)) {
+					drv_data->gyro_x =
+						((drv_data->fifo_data[1] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[2] &
+						   0xff);
+					drv_data->gyro_y =
+						((drv_data->fifo_data[3] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[4] &
+						   0xff);
+					drv_data->gyro_z =
+						((drv_data->fifo_data[5] &
+						  0xff) << 8)
+						+ (drv_data->fifo_data[6] &
+						   0xff);
+				}
+				drv_data->temp =
+					(int16_t)(drv_data->fifo_data[7]);
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int icm42605_attr_set(struct device *dev,
+			     enum sensor_channel chan,
+			     enum sensor_attribute attr,
+			     const struct sensor_value *val)
+{
+	struct icm42605_data *drv_data = dev->data;
+
+	__ASSERT_NO_MSG(val != NULL);
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		if (attr == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+			if (val->val1 > 1000 || val->val1 < 12) {
+				LOG_ERR("Incorrect sampling value");
+				return -EINVAL;
+			} else {
+				drv_data->accel_hz = val->val1;
+			}
+		} else if (attr == SENSOR_ATTR_FULL_SCALE) {
+			if (val->val1 < ACCEL_FS_16G ||
+			    val->val1 > ACCEL_FS_2G) {
+				LOG_ERR("Incorrect fullscale value");
+				return -EINVAL;
+			} else {
+				drv_data->accel_sf = val->val1;
+			}
+		} else {
+			LOG_ERR("Not supported ATTR");
+			return -ENOTSUP;
+		}
+
+		break;
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+	case SENSOR_CHAN_GYRO_XYZ:
+		if (attr == SENSOR_ATTR_SAMPLING_FREQUENCY) {
+			if (val->val1 > 1000 || val->val1 < 12) {
+				LOG_ERR("Incorrect sampling value");
+				return -EINVAL;
+			} else {
+				drv_data->gyro_hz = val->val1;
+			}
+		} else if (attr == SENSOR_ATTR_FULL_SCALE) {
+			if (val->val1 < GYRO_FS_2000DPS ||
```

Comment:
Attributes are passed and returned with units (see https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/sensor.h#L266). So for example, for sampling frequency you could get "3.125" (val1 = 3, val2 = 125000). So using a chain of if/else if/else you should map that into your internal enum/definition. Only supported values should be handled.

Reply:
@gmarull I think floating point based frequency make confusion to user.
That is why Invensense doesn't use floating point value for Frequency even current sensor provide floating point value at low frequency.
If Zephyer sensor driver should use floating point based frequency, I will implement.

Reply:
@gmarull Misunderstand.. I will fix it.

Reply:
@gmarull Is there any way to get float value from devicetree?
For use floating point value of HZ and FS, need to get float value from devicetree.
Or I need to separate accel_hz or accel_fs to accel_hz_int, accel_hz_frac / accel_fs_int, accel_fs_frac
Need any idea about this implementation..

                .accel_hz = DT_INST_PROP(index, accel_hz),              \
                .gyro_hz = DT_INST_PROP(index, gyro_hz),                \
                .accel_fs = DT_ENUM_IDX(DT_DRV_INST(index), accel_fs),  \
                .gyro_fs = DT_ENUM_IDX(DT_DRV_INST(index), gyro_fs),    \

Reply:
No, floats are not supported on DT. Solutions I can think of is either use another unit scale (for example if it was voltage mV instead of V), or, use string based enums that then translate to appropriate values.

Reply:
@JHKim8212 Floats are not used by the sensor API anyway, the sensor_value struct is used where you have the integer part and the decimal part both as integers.

Reply:
@gmarull Can sensor driver use enum value for sensor frequency instead of actual value with int and fractional?
Code is too heavy like below..


        if (a_rate.val1 >= 4000) {
                if (a_rate.val1 == 4000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_4000;
                } else {
                        databuf |= BIT_ACCEL_ODR_8000;
                }
        } else if (a_rate.val1 >= 2000) {
                if (a_rate.val1 == 2000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_2000;
                } else {
                        databuf |= BIT_ACCEL_ODR_4000;
                }
        } else if (a_rate.val1 >= 1000) {
                if (a_rate.val1 == 1000 && a_rate.val2 <= 0) {
                        databuf |= BIT_ACCEL_ODR_1000;
                } else {
                        databuf |= BIT_ACCEL_ODR_2000;
                }

Reply:
@gmarull Can you please give me opinion about change actual value of Hz to ENUM?